### PR TITLE
[Update categorical.py] Creating `Categorical` distribution much faster

### LINKED ISSUE #11747: [distributions]: Creating `Categorical` is 100x slower than `Normal`
In RL, we often need to create a distribution over actions at each time step during rollout by using the policy network to select actions. It seems now creating `Categorical` distribution is much slower than `Normal` distribution during rollout. 

## Toy code:

```python
from time import time
import torch
import torch.nn.functional as F
from torch.distributions import Categorical, Normal


t = time()
for _ in range(100):
    probs = F.softmax(torch.randn(1, 32), -1)
    dist = Categorical(probs)
t1 = time() - t
print(t1)

t = time()
for _ in range(100):
    mu = torch.randn(1, 32)
    std = torch.randn(1, 32)
    dist = Normal(mu, std)
t2 = time() - t
print(t2)

print(t1/t2)
```
gives 
```
0.5241701602935791
0.002852201461791992
183.77739697400318
```
which almost 180x slower for `Categorical`. 

By tracing back to the code, it turns out the slow part is contributed by [this line](https://github.com/pytorch/pytorch/blob/master/torch/distributions/categorical.py#L50), another toy code to show this

```python
t = time()
for _ in range(100):
    p = F.softmax(torch.randn(1, 32), -1)
    p = p/p.sum(-1, keepdim=True)
print(time() - t)
```
gives
```
0.4742550849914551
```

## System Info

PyTorch: 0.4.1

### PR DESCRIPTION
**Pull request by zuoxingdong**
Fix #11747

### CODE REVIEW

File: torch/distributions/categorical.py [hunk 1]
```diff
@@ -47,7 +47,9 @@ def __init__(self, probs=None, logits=None, validate_args=None):
         if probs is not None:
             if probs.dim() < 1:
                 raise ValueError("`probs` parameter must be at least one-dimensional.")
-            self.probs = probs / probs.sum(-1, keepdim=True)
+            sum_probs = probs.detach().cpu().numpy().sum(-1, keepdims=True)
```
Comment (vishwakftw) [inline torch/distributions/categorical.py]: This places a dependency on NumPy unnecessarily.

Comment (zuoxingdong) [reply torch/distributions/categorical.py]: @vishwakftw `.numpy()` will returns a numpy array, but we do not explicitly require `import numpy` in this file. Is this what you mean ?

Comment (ezyang) [reply torch/distributions/categorical.py]: `numpy()` doesn't work unless you have a copy of PyTorch which was compiled with Numpy support. We don't want to assume that this is the case in core code.

Comment (zuoxingdong) [reply torch/distributions/categorical.py]: @ezyang Thanks for the comment. In this case, I would think to remove the 're-normalize' operation and let user to take responsibility for the validity of probabilities. Because with this line, it is really really slow in certain cases e.g. I am training policy with ES to solve RL tasks, the discrete (`Categorical`) environment is extremely slow compared with continuous tasks (`Normal`), because we have to create distribution for each time step. In addition, before creating categorical distribution, the most common use cases are either to take a softmax and feed probabilities in or feed the logits in. Thus, I would vote for removing this line of code as the violation of validity is rare but the degrading of speed is very high. What do you think ?

Comment (vishwakftw) [reply torch/distributions/categorical.py]: I did benchmark the operation: sum along the last dim and add, and it seems to be slower on NumPy actually.
With PyTorch (cpu):
`33.2 ms ± 1.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`
With PyTorch (gpu, 1080Ti):
`3.74 ms ± 28.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`
With NumPy:
`189 ms ± 8.33 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`

And the `detach` is breaking autograd tests.

Comment (zuoxingdong) [reply torch/distributions/categorical.py]: @vishwakftw I made the following test case

```
from time import time
import numpy as np
import torch
import torch.nn.functional as F

t = time()
for _ in range(100):
    p = F.softmax(torch.randn(1, 32), -1)
    p = p/p.sum(-1, keepdim=True)
print(time() - t)

t = time()
for _ in range(100):
    p = F.softmax(torch.randn(1, 32), -1)
    
    p = p/torch.from_numpy(p.detach().numpy().sum(-1, keepdims=True)).float()
print(time() - t)
```
the Numpy alternative is much faster.

But anyway as @ezyang pointed out above, numpy should not be considered here. 

The original motivation is in #11747 . Do you have some idea how to reimplement it without using `.sum()` function ?

Comment (alicanb) [reply torch/distributions/categorical.py]: Can't reproduce those times btw, running your code gives me 
`0.007228851318359375
0.007979869842529297` with pytorch 0.4.1, on google colab no gpu

Comment (zuoxingdong) [reply torch/distributions/categorical.py]: @alicanb @vishwakftw Indeed. The main use case I am referring to is using ES to evolve a small policy to solve sequential problems (RL), which requires a lot of samples each over a long horizon, this rises of the need being able to create thousands of 'small' categorical distributions efficiently. On the ES code I tested on, with same setting, it becomes extremely slow compared with other distributions like `Normal` for continuous control tasks.

**Review by apaszke — CHANGES_REQUESTED** [review]
Breaks when PyTorch is compiled without NumPy support, forces a GPU sync point, and breaks differentiation

**Comment by zuoxingdong** [comment]
@apaszke At present, I would suggest to remove this normalization check and let the user takes the responsibility to ensure the `probs` is a valid probabilities. Despite of the pro of being much faster, and the potential con of being violated as a probability vector is rather rare since most use cases come from applying softmax before creating categorical distribution. What do you think ?

**Comment by fmassa** [comment]
We will soon be update g our reduction code to be more efficient, so I'm not sure we need this

**Comment by zuoxingdong** [comment]
@fmassa If I got it correctly, you mean the function like `sum()` will be rewritten by using `TensorIterator` which would be much faster ? 

Currently, [this line](https://github.com/pytorch/pytorch/blob/master/torch/distributions/categorical.py#L50) also makes my policy gradient evaluation phase extremely slow on discrete control tasks compared with continuous ones (with `Normal`)

**Comment by zuoxingdong** [comment]
@fmassa Since the API already indicates that the argument `probs` should be valid probabilities, I am wondering why we need to do the normalization as most of the cases the user would already get a `softmax` operator beforehand.

**Comment by fmassa** [comment]
@zuoxingdong yes, `sum` is being rewritten using a completely different strategy. The current WIP branch is in https://github.com/colesbury/pytorch/tree/tensor_iterator_sum

**Comment by SsnL** [comment]
I agree with @fmassa .I think the correct fix would be improve the sum operator rather than using numpy to do the calculation here. @colesbury 's changes on `sum()` operator should be able to speed this up a lot.

**Comment by t-vi** [comment]
Given that this has been dormant for over a year and we have TensorIterator sum for a long time, is this still needed?
